File: srcs/juloo.keyboard2/BackupRestoreManager.java
package juloo.keyboard2;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.util.DisplayMetrics;
import android.util.Log;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashSet;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

/**
 * Manages backup and restore of keyboard configuration
 * Uses Storage Access Framework (SAF) for Android 15+ compatibility
 */
public class BackupRestoreManager
{
  private static final String TAG = "BackupRestoreManager";
  private final Context context;
  private final Gson gson;

  public BackupRestoreManager(Context context)
  {
    this.context = context;
    this.gson = new GsonBuilder().setPrettyPrinting().create();
  }

  /**
   * Export all preferences to JSON file
   * @param uri URI from Storage Access Framework (ACTION_CREATE_DOCUMENT)
   * @return true if successful
   */
  public boolean exportConfig(Uri uri, SharedPreferences prefs) throws Exception
  {
    try
    {
      // Collect metadata
      JsonObject root = new JsonObject();
      JsonObject metadata = new JsonObject();

      // App version
      String versionName = context.getPackageManager()
        .getPackageInfo(context.getPackageName(), 0).versionName;
      int versionCode = context.getPackageManager()
        .getPackageInfo(context.getPackageName(), 0).versionCode;

      metadata.addProperty("app_version", versionName);
      metadata.addProperty("version_code", versionCode);
      metadata.addProperty("export_date",
        new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US).format(new Date()));

      // Screen dimensions
      DisplayMetrics dm = context.getResources().getDisplayMetrics();
      metadata.addProperty("screen_width", dm.widthPixels);
      metadata.addProperty("screen_height", dm.heightPixels);
      metadata.addProperty("screen_density", dm.density);
      metadata.addProperty("android_version", android.os.Build.VERSION.SDK_INT);

      root.add("metadata", metadata);

      // Export all preferences
      Map<String, ?> allPrefs = prefs.getAll();
      JsonObject preferences = new JsonObject();

      for (Map.Entry<String, ?> entry : allPrefs.entrySet())
      {
        String key = entry.getKey();
        Object value = entry.getValue();

        // Preserve JSON-string preferences (layouts, extra_keys, custom_extra_keys)
        // These are already stored as JSON strings and should be preserved as-is
        if (isJsonStringPreference(key) && value instanceof String)
        {
          try
          {
            // Parse the JSON string and add as JsonElement to avoid double-encoding
            preferences.add(key, JsonParser.parseString((String)value));
          }
          catch (Exception e)
          {
            Log.w(TAG, "Failed to parse JSON preference: " + key, e);
            // Fall back to regular serialization if parsing fails
            preferences.add(key, gson.toJsonTree(value));
          }
        }
        else if (isInternalPreference(key))
        {
          // Skip internal state preferences
          Log.i(TAG, "Skipping internal preference on export: " + key);
        }
        else
        {
          preferences.add(key, gson.toJsonTree(value));
        }
      }

      root.add("preferences", preferences);

      // Write to file
      try (OutputStream outputStream = context.getContentResolver().openOutputStream(uri);
           OutputStreamWriter writer = new OutputStreamWriter(outputStream))
      {
        gson.toJson(root, writer);
        writer.flush();
      }

      Log.i(TAG, "Exported " + allPrefs.size() + " preferences");
      return true;
    }
    catch (Exception e)
    {
      Log.e(TAG, "Export failed", e);
      throw new Exception("Export failed: " + e.getMessage(), e);
    }
  }

  /**
   * Import preferences from JSON file with version-tolerant parsing
   * @param uri URI from Storage Access Framework (ACTION_OPEN_DOCUMENT)
   * @return ImportResult with statistics
   */
  public ImportResult importConfig(Uri uri, SharedPreferences prefs) throws Exception
  {
    try
    {
      // Read JSON file
      StringBuilder jsonBuilder = new StringBuilder();
      try (BufferedReader reader = new BufferedReader(
        new InputStreamReader(context.getContentResolver().openInputStream(uri))))
      {
        String line;
        while ((line = reader.readLine()) != null)
        {
          jsonBuilder.append(line);
        }
      }

      JsonObject root = JsonParser.parseString(jsonBuilder.toString()).getAsJsonObject();

      // Parse metadata (optional, for informational purposes)
      ImportResult result = new ImportResult();
      if (root.has("metadata"))
      {
        JsonObject metadata = root.getAsJsonObject("metadata");
        result.sourceVersion = metadata.has("app_version") ?
          metadata.get("app_version").getAsString() : "unknown";
        result.sourceScreenWidth = metadata.has("screen_width") ?
          metadata.get("screen_width").getAsInt() : 0;
        result.sourceScreenHeight = metadata.has("screen_height") ?
          metadata.get("screen_height").getAsInt() : 0;
      }

      // Get current screen dimensions for comparison
      DisplayMetrics dm = context.getResources().getDisplayMetrics();
      result.currentScreenWidth = dm.widthPixels;
      result.currentScreenHeight = dm.heightPixels;

      // Import preferences with validation
      if (!root.has("preferences"))
      {
        throw new Exception("Invalid backup file: missing preferences section");
      }

      JsonObject preferences = root.getAsJsonObject("preferences");
      SharedPreferences.Editor editor = prefs.edit();

      int imported = 0;
      int skipped = 0;

      for (Map.Entry<String, com.google.gson.JsonElement> entry : preferences.entrySet())
      {
        String key = entry.getKey();
        com.google.gson.JsonElement value = entry.getValue();

        try
        {
          if (importPreference(editor, key, value))
          {
            imported++;
            result.importedKeys.add(key);
          }
          else
          {
            skipped++;
            result.skippedKeys.add(key);
          }
        }
        catch (Exception e)
        {
          Log.w(TAG, "Failed to import key: " + key, e);
          skipped++;
          result.skippedKeys.add(key);
        }
      }

      editor.apply();

      result.importedCount = imported;
      result.skippedCount = skipped;

      Log.i(TAG, "Import complete: " + imported + " imported, " + skipped + " skipped");
      return result;
    }
    catch (Exception e)
    {
      Log.e(TAG, "Import failed", e);
      throw new Exception("Import failed: " + e.getMessage(), e);
    }
  }

  /**
   * Import a single preference with type detection and validation
   * @return true if imported, false if skipped
   */
  private boolean importPreference(SharedPreferences.Editor editor, String key,
                                    com.google.gson.JsonElement value)
  {
    // Skip internal state preferences
    if (isInternalPreference(key))
    {
      Log.i(TAG, "Skipping internal preference: " + key);
      return false;
    }

    // Handle JSON-string preferences (layouts, extra_keys, custom_extra_keys)
    // These are stored as JSON strings in SharedPreferences
    if (isJsonStringPreference(key))
    {
      String jsonString;

      // Handle both old double-encoded strings and new native JSON arrays/objects
      if (value.isJsonPrimitive() && value.getAsJsonPrimitive().isString())
      {
        // Old format: the JSON was exported as a string primitive (double-encoded)
        // Just use the string value directly
        jsonString = value.getAsString();
        Log.i(TAG, "Importing old-format JSON-string preference: " + key);
      }
      else if (value.isJsonArray() || value.isJsonObject())
      {
        // New format: the JSON is a native array/object
        // Convert to compact JSON string (no pretty printing for preferences)
        jsonString = value.toString();
        Log.i(TAG, "Importing new-format JSON-string preference: " + key);
      }
      else
      {
        Log.w(TAG, "Unexpected format for JSON-string preference: " + key);
        return false;
      }

      editor.putString(key, jsonString);
      return true;
    }
    // Handle different preference types
    if (value.isJsonPrimitive())
    {
      com.google.gson.JsonPrimitive primitive = value.getAsJsonPrimitive();

      if (primitive.isBoolean())
      {
        editor.putBoolean(key, primitive.getAsBoolean());
        return true;
      }
      else if (primitive.isNumber())
      {
        // Check if this preference is known to be a float type
        // We must use the correct type because SharedPreferences throws ClassCastException
        // if we try to read an int as float or vice versa
        if (isFloatPreference(key))
        {
          // Known float preference
          float floatValue = primitive.getAsFloat();
          if (validateFloatPreference(key, floatValue))
          {
            editor.putFloat(key, floatValue);
            return true;
          }
          else
          {
            Log.w(TAG, "Skipping invalid float value for " + key + ": " + floatValue);
            return false;
          }
        }
        else
        {
          // Assume integer for all other numeric preferences
          int intValue = primitive.getAsInt();
          if (validateIntPreference(key, intValue))
          {
            editor.putInt(key, intValue);
            return true;
          }
          else
          {
            Log.w(TAG, "Skipping invalid int value for " + key + ": " + intValue);
            return false;
          }
        }
      }
      else if (primitive.isString())
      {
        String stringValue = primitive.getAsString();

        // Some preferences store integers as strings (from ListPreference)
        // Parse and store them as actual integers to prevent ClassCastException
        if (isIntegerStoredAsString(key))
        {
          try
          {
            int intValue = Integer.parseInt(stringValue);
            if (validateIntPreference(key, intValue))
            {
              editor.putInt(key, intValue);
              return true;
            }
            else
            {
              Log.w(TAG, "Skipping invalid int-as-string value for " + key + ": " + intValue);
              return false;
            }
          }
          catch (NumberFormatException e)
          {
            Log.w(TAG, "Failed to parse int-as-string for " + key + ": " + stringValue);
            return false;
          }
        }

        if (validateStringPreference(key, stringValue))
        {
          editor.putString(key, stringValue);
          return true;
        }
        else
        {
          Log.w(TAG, "Skipping invalid string value for " + key);
          return false;
        }
      }
    }
    else if (value.isJsonArray())
    {
      // Handle String Sets
      Set<String> stringSet = new HashSet<>();
      for (com.google.gson.JsonElement element : value.getAsJsonArray())
      {
        if (element.isJsonPrimitive() && element.getAsJsonPrimitive().isString())
        {
          stringSet.add(element.getAsString());
        }
      }
      editor.putStringSet(key, stringSet);
      return true;
    }

    Log.w(TAG, "Skipping unknown preference type for " + key);
    return false;
  }

  /**
   * Validate integer preference values
   */
  private boolean validateIntPreference(String key, int value)
  {
    switch (key)
    {
      // Opacity values (0-100)
      case "label_brightness":
      case "keyboard_opacity":
      case "key_opacity":
      case "key_activated_opacity":
      case "suggestion_bar_opacity":
        return value >= 0 && value <= 100;

      // Keyboard height percentages
      case "keyboard_height":
      case "keyboard_height_unfolded":
        return value >= 10 && value <= 100;
      case "keyboard_height_landscape":
      case "keyboard_height_landscape_unfolded":
        return value >= 20 && value <= 65;

      // Margins and spacing (0-200 dp max)
      case "margin_bottom_portrait":
      case "margin_bottom_landscape":
      case "margin_bottom_portrait_unfolded":
      case "margin_bottom_landscape_unfolded":
      case "horizontal_margin_portrait":
      case "horizontal_margin_landscape":
      case "horizontal_margin_portrait_unfolded":
      case "horizontal_margin_landscape_unfolded":
        return value >= 0 && value <= 200;

      // Border radius (0-100%)
      case "custom_border_radius":
        return value >= 0 && value <= 100;

      // Timing values (milliseconds)
      case "vibrate_duration":
        return value >= 0 && value <= 100;
      case "longpress_timeout":
        return value >= 50 && value <= 2000;
      case "longpress_interval":
        return value >= 5 && value <= 100;

      // Short gesture distance (10-95%)
      case "short_gesture_min_distance":
        return value >= 10 && value <= 95;

      // Neural network parameters
      case "neural_beam_width":
        return value >= 1 && value <= 16;
      case "neural_max_length":
        return value >= 10 && value <= 50;

      // Auto-correction parameters
      case "autocorrect_min_word_length":
        return value >= 2 && value <= 5;
      case "autocorrect_confidence_min_frequency":
        return value >= 100 && value <= 5000;

      // Clipboard history limit
      case "clipboard_history_limit":
        return value >= 1 && value <= 50;

      // Circle sensitivity
      case "circle_sensitivity":
        return value >= 1 && value <= 5;

      default:
        // Unknown integer preference - allow it (version-tolerant)
        return true;
    }
  }

  /**
   * Validate float preference values
   */
  private boolean validateFloatPreference(String key, float value)
  {
    switch (key)
    {
      // Character size (0.75-1.5)
      case "character_size":
        return value >= 0.75f && value <= 1.5f;

      // Margins (0-5%)
      case "key_vertical_margin":
      case "key_horizontal_margin":
        return value >= 0f && value <= 5f;

      // Border line width (0-5 dp)
      case "custom_border_line_width":
        return value >= 0f && value <= 5f;

      // Prediction weights
      case "prediction_context_boost":
        return value >= 0.5f && value <= 5.0f;
      case "prediction_frequency_scale":
        return value >= 100.0f && value <= 5000.0f;

      // Auto-correction threshold
      case "autocorrect_char_match_threshold":
        return value >= 0.5f && value <= 0.9f;

      // Neural confidence threshold
      case "neural_confidence_threshold":
        return value >= 0.0f && value <= 1.0f;

      default:
        // Unknown float preference - allow it (version-tolerant)
        return true;
    }
  }

  /**
   * Check if a preference stores data as a JSON string
   * These preferences use ListGroupPreference which stores data as JSON-encoded strings
   */
  private boolean isJsonStringPreference(String key)
  {
    switch (key)
    {
      // LayoutsPreference - stores List<Layout> as JSON string
      case "layouts":
      // ExtraKeysPreference - stores Map<KeyValue, PreferredPos> as JSON string
      case "extra_keys":
      // CustomExtraKeysPreference - stores Map<KeyValue, PreferredPos> as JSON string
      case "custom_extra_keys":
        return true;
      default:
        return false;
    }
  }

  /**
   * Check if a preference is internal state that shouldn't be exported/imported
   */
  private boolean isInternalPreference(String key)
  {
    switch (key)
    {
      // Internal version tracking
      case "version":
      // Current layout indices (managed by Config, device-specific)
      case "current_layout_portrait":
      case "current_layout_landscape":
        return true;
      default:
        return false;
    }
  }

  /**
   * Check if a preference is stored as a float in SharedPreferences
   * This is critical because SharedPreferences throws ClassCastException if you
   * try to read an int as float or vice versa
   */
  private boolean isFloatPreference(String key)
  {
    switch (key)
    {
      // Character and UI sizing
      case "character_size":
      case "key_vertical_margin":
      case "key_horizontal_margin":
      case "custom_border_line_width":

      // Prediction weights
      case "prediction_context_boost":
      case "prediction_frequency_scale":

      // Auto-correction threshold
      case "autocorrect_char_match_threshold":

      // Neural confidence threshold
      case "neural_confidence_threshold":
        return true;
      default:
        return false;
    }
  }

  /**
   * Check if a preference stores integers as strings (from ListPreference)
   * These need to be parsed and stored as int to prevent ClassCastException
   */
  private boolean isIntegerStoredAsString(String key)
  {
    switch (key)
    {
      // ListPreference values that are actually integers
      case "circle_sensitivity":
      case "show_numpad":
      case "clipboard_history_limit":
        return true;
      default:
        return false;
    }
  }

  /**
   * Validate string preference values
   */
  private boolean validateStringPreference(String key, String value)
  {
    if (value == null)
      return false;

    switch (key)
    {
      // Theme values - relaxed validation for forward compatibility
      // New themes added in future versions should still import successfully
      case "theme":
        // Just ensure it's not empty - app will fall back to default if invalid
        return !value.isEmpty();

      // Number row options
      case "number_row":
        return value.matches("no_number_row|no_symbols|symbols");

      // Show numpad options
      case "show_numpad":
        return value.matches("never|always|landscape|[0-9]+");

      // Numpad layout
      case "numpad_layout":
        return value.matches("high_first|low_first|default");

      // Number entry layout
      case "number_entry_layout":
        return value.matches("pin|number");

      // Circle sensitivity (string representation)
      case "circle_sensitivity":
        return value.matches("[1-5]");

      // Slider sensitivity (string representation)
      case "slider_sensitivity":
        return value.matches("[0-9]+");

      // Swipe distance (string representation)
      case "swipe_dist":
        return value.matches("[0-9]+(\\.[0-9]+)?");

      default:
        // Unknown string preference - allow it (version-tolerant)
        return true;
    }
  }

  /**
   * Result of import operation
   */
  public static class ImportResult
  {
    public int importedCount = 0;
    public int skippedCount = 0;
    public String sourceVersion = "unknown";
    public int sourceScreenWidth = 0;
    public int sourceScreenHeight = 0;
    public int currentScreenWidth = 0;
    public int currentScreenHeight = 0;
    public Set<String> importedKeys = new HashSet<>();
    public Set<String> skippedKeys = new HashSet<>();

    public boolean hasScreenSizeMismatch()
    {
      if (sourceScreenWidth == 0 || sourceScreenHeight == 0)
        return false; // No source dimensions available

      int widthDiff = Math.abs(currentScreenWidth - sourceScreenWidth);
      int heightDiff = Math.abs(currentScreenHeight - sourceScreenHeight);

      // Consider it a mismatch if either dimension differs by more than 20%
      return (widthDiff > currentScreenWidth * 0.2) ||
             (heightDiff > currentScreenHeight * 0.2);
    }
  }
}

File: srcs/juloo.keyboard2/BackupRestoreManager.kt
package juloo.keyboard2

import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import android.util.Log
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.io.BufferedReader
import java.io.InputStreamReader
import java.text.SimpleDateFormat
import java.util.*
import kotlin.math.abs

/**
 * Manages backup and restore of keyboard configuration
 * Uses Storage Access Framework (SAF) for Android 15+ compatibility
 */
class BackupRestoreManager(private val context: Context) {
    private val gson: Gson = GsonBuilder().setPrettyPrinting().create()

    /**
     * Export all preferences to JSON file
     * @param uri URI from Storage Access Framework (ACTION_CREATE_DOCUMENT)
     * @return true if successful
     */
    fun exportConfig(uri: Uri, prefs: SharedPreferences): Boolean {
        return try {
            // Collect metadata
            val root = JsonObject()
            val metadata = JsonObject()

            // App version
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            val versionName = packageInfo.versionName
            val versionCode = packageInfo.versionCode

            metadata.addProperty("app_version", versionName)
            metadata.addProperty("version_code", versionCode)
            metadata.addProperty(
                "export_date",
                SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US).format(Date())
            )

            // Screen dimensions
            val dm = context.resources.displayMetrics
            metadata.addProperty("screen_width", dm.widthPixels)
            metadata.addProperty("screen_height", dm.heightPixels)
            metadata.addProperty("screen_density", dm.density)
            metadata.addProperty("android_version", android.os.Build.VERSION.SDK_INT)

            root.add("metadata", metadata)

            // Export all preferences
            val allPrefs = prefs.all
            val preferences = JsonObject()

            for ((key, value) in allPrefs) {
                // Preserve JSON-string preferences (layouts, extra_keys, custom_extra_keys)
                // These are already stored as JSON strings and should be preserved as-is
                when {
                    isJsonStringPreference(key) && value is String -> {
                        try {
                            // Parse the JSON string and add as JsonElement to avoid double-encoding
                            preferences.add(key, JsonParser.parseString(value))
                        } catch (e: Exception) {
                            Log.w(TAG, "Failed to parse JSON preference: $key", e)
                            // Fall back to regular serialization if parsing fails
                            preferences.add(key, gson.toJsonTree(value))
                        }
                    }
                    isInternalPreference(key) -> {
                        // Skip internal state preferences
                        Log.i(TAG, "Skipping internal preference on export: $key")
                    }
                    else -> {
                        preferences.add(key, gson.toJsonTree(value))
                    }
                }
            }

            root.add("preferences", preferences)

            // Write to file
            context.contentResolver.openOutputStream(uri)?.use { outputStream ->
                outputStream.writer().use { writer ->
                    gson.toJson(root, writer)
                    writer.flush()
                }
            }

            Log.i(TAG, "Exported ${preferences.size()} preferences (out of ${allPrefs.size} total)")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Export failed", e)
            throw Exception("Export failed: ${e.message}", e)
        }
    }

    /**
     * Import preferences from JSON file with version-tolerant parsing
     * @param uri URI from Storage Access Framework (ACTION_OPEN_DOCUMENT)
     * @return ImportResult with statistics
     */
    fun importConfig(uri: Uri, prefs: SharedPreferences): ImportResult {
        return try {
            // Read JSON file
            val jsonBuilder = StringBuilder()
            context.contentResolver.openInputStream(uri)?.use { inputStream ->
                BufferedReader(InputStreamReader(inputStream)).use { reader ->
                    reader.forEachLine { line ->
                        jsonBuilder.append(line)
                    }
                }
            }

            val root = JsonParser.parseString(jsonBuilder.toString()).asJsonObject

            // Parse metadata (optional, for informational purposes)
            val result = ImportResult()
            if (root.has("metadata")) {
                val metadata = root.getAsJsonObject("metadata")
                result.sourceVersion = metadata.get("app_version")?.asString ?: "unknown"
                result.sourceScreenWidth = metadata.get("screen_width")?.asInt ?: 0
                result.sourceScreenHeight = metadata.get("screen_height")?.asInt ?: 0
            }

            // Get current screen dimensions for comparison
            val dm = context.resources.displayMetrics
            result.currentScreenWidth = dm.widthPixels
            result.currentScreenHeight = dm.heightPixels

            // Import preferences with validation
            if (!root.has("preferences")) {
                throw Exception("Invalid backup file: missing preferences section")
            }

            val preferences = root.getAsJsonObject("preferences")
            val editor = prefs.edit()

            var imported = 0
            var skipped = 0

            for ((key, value) in preferences.entrySet()) {
                try {
                    if (importPreference(editor, key, value)) {
                        imported++
                        result.importedKeys.add(key)
                        Log.d(TAG, "Imported: $key = $value")
                    } else {
                        skipped++
                        result.skippedKeys.add(key)
                        Log.i(TAG, "Skipped: $key = $value")
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to import key: $key = $value", e)
                    skipped++
                    result.skippedKeys.add(key)
                }
            }

            editor.apply()

            result.importedCount = imported
            result.skippedCount = skipped

            Log.i(TAG, "Import complete: $imported imported, $skipped skipped")
            result
        } catch (e: Exception) {
            Log.e(TAG, "Import failed", e)
            throw Exception("Import failed: ${e.message}", e)
        }
    }

    /**
     * Import a single preference with type detection and validation
     * @return true if imported, false if skipped
     */
    private fun importPreference(editor: SharedPreferences.Editor, key: String, value: JsonElement): Boolean {
        // Skip internal state preferences
        if (isInternalPreference(key)) {
            Log.i(TAG, "Skipping internal preference: $key")
            return false
        }

        // Handle JSON-string preferences (layouts, extra_keys, custom_extra_keys)
        // These are stored as JSON strings in SharedPreferences
        if (isJsonStringPreference(key)) {
            val jsonString = when {
                // Old format: the JSON was exported as a string primitive (double-encoded)
                value.isJsonPrimitive && value.asJsonPrimitive.isString -> {
                    Log.i(TAG, "Importing old-format JSON-string preference: $key")
                    value.asString
                }
                // New format: the JSON is a native array/object
                value.isJsonArray || value.isJsonObject -> {
                    Log.i(TAG, "Importing new-format JSON-string preference: $key")
                    value.toString()
                }
                else -> {
                    Log.w(TAG, "Unexpected format for JSON-string preference: $key")
                    return false
                }
            }

            editor.putString(key, jsonString)
            return true
        }

        // Handle different preference types
        when {
            value.isJsonPrimitive -> {
                val primitive = value.asJsonPrimitive

                when {
                    primitive.isBoolean -> {
                        editor.putBoolean(key, primitive.asBoolean)
                        return true
                    }
                    primitive.isNumber -> {
                        // Check if this preference is known to be a float type
                        if (isFloatPreference(key)) {
                            val floatValue = primitive.asFloat
                            if (validateFloatPreference(key, floatValue)) {
                                editor.putFloat(key, floatValue)
                                return true
                            } else {
                                Log.w(TAG, "Skipping invalid float value for $key: $floatValue")
                                return false
                            }
                        } else {
                            // Assume integer for all other numeric preferences
                            val intValue = primitive.asInt
                            if (validateIntPreference(key, intValue)) {
                                editor.putInt(key, intValue)
                                return true
                            } else {
                                Log.w(TAG, "Skipping invalid int value for $key: $intValue")
                                return false
                            }
                        }
                    }
                    primitive.isString -> {
                        val stringValue = primitive.asString

                        // Some preferences store integers as strings (from ListPreference)
                        if (isIntegerStoredAsString(key)) {
                            try {
                                val intValue = stringValue.toInt()
                                if (validateIntPreference(key, intValue)) {
                                    editor.putInt(key, intValue)
                                    return true
                                } else {
                                    Log.w(TAG, "Skipping invalid int-as-string value for $key: $intValue")
                                    return false
                                }
                            } catch (e: NumberFormatException) {
                                Log.w(TAG, "Failed to parse int-as-string for $key: $stringValue")
                                return false
                            }
                        }

                        if (validateStringPreference(key, stringValue)) {
                            editor.putString(key, stringValue)
                            return true
                        } else {
                            Log.w(TAG, "Skipping invalid string value for $key")
                            return false
                        }
                    }
                }
            }
            value.isJsonArray -> {
                // Only parse as StringSet if this preference is known to be a StringSet
                if (isStringSetPreference(key)) {
                    val stringSet = mutableSetOf<String>()
                    for (element in value.asJsonArray) {
                        if (element.isJsonPrimitive && element.asJsonPrimitive.isString) {
                            stringSet.add(element.asString)
                        }
                    }
                    editor.putStringSet(key, stringSet)
                    return true
                } else {
                    Log.w(TAG, "Skipping unexpected JsonArray for key: $key")
                    return false
                }
            }
            value.isJsonNull -> {
                // Null values - skip (can't store null in SharedPreferences)
                Log.i(TAG, "Skipping null preference: $key")
                return false
            }
            value.isJsonObject -> {
                // Unexpected JsonObject that's not a JSON-string preference
                Log.w(TAG, "Skipping unexpected JsonObject preference: $key")
                return false
            }
        }

        Log.w(TAG, "Skipping unknown preference type for $key type=${value.javaClass.simpleName}")
        return false
    }

    /**
     * Validate integer preference values
     */
    private fun validateIntPreference(key: String, value: Int): Boolean {
        return when (key) {
            // Opacity values (0-100)
            "label_brightness", "keyboard_opacity", "key_opacity",
            "key_activated_opacity", "suggestion_bar_opacity" -> value in 0..100

            // Keyboard height percentages
            "keyboard_height", "keyboard_height_unfolded" -> value in 10..100
            "keyboard_height_landscape", "keyboard_height_landscape_unfolded" -> value in 20..65

            // Margins and spacing (0-200 dp max)
            "margin_bottom_portrait", "margin_bottom_landscape",
            "margin_bottom_portrait_unfolded", "margin_bottom_landscape_unfolded",
            "horizontal_margin_portrait", "horizontal_margin_landscape",
            "horizontal_margin_portrait_unfolded", "horizontal_margin_landscape_unfolded" -> value in 0..200

            // Border radius (0-100%)
            "custom_border_radius" -> value in 0..100

            // Timing values (milliseconds)
            "vibrate_duration" -> value in 0..100
            "longpress_timeout" -> value in 50..2000
            "longpress_interval" -> value in 5..100

            // Short gesture distance (10-95%)
            "short_gesture_min_distance" -> value in 10..95

            // Neural network parameters
            "neural_beam_width" -> value in 1..16
            "neural_max_length" -> value in 10..50

            // Auto-correction parameters
            "autocorrect_min_word_length" -> value in 2..5
            "autocorrect_confidence_min_frequency" -> value in 100..5000

            // Clipboard history limit
            "clipboard_history_limit" -> value in 1..50

            // Circle sensitivity
            "circle_sensitivity" -> value in 1..5

            // Unknown integer preference - allow it (version-tolerant)
            else -> true
        }
    }

    /**
     * Validate float preference values
     */
    private fun validateFloatPreference(key: String, value: Float): Boolean {
        return when (key) {
            // Character size (0.75-1.5)
            "character_size" -> value in 0.75f..1.5f

            // Margins (0-5%)
            "key_vertical_margin", "key_horizontal_margin" -> value in 0f..5f

            // Border line width (0-5 dp)
            "custom_border_line_width" -> value in 0f..5f

            // Prediction weights
            "prediction_context_boost" -> value in 0.5f..5.0f
            "prediction_frequency_scale" -> value in 100.0f..5000.0f

            // Auto-correction threshold
            "autocorrect_char_match_threshold" -> value in 0.5f..0.9f

            // Neural confidence threshold
            "neural_confidence_threshold" -> value in 0.0f..1.0f

            // Swipe typing boost parameters (0.0-2.0 range)
            "swipe_rare_words_penalty", "swipe_common_words_boost", "swipe_top5000_boost" -> value in 0.0f..2.0f

            // Unknown float preference - allow it (version-tolerant)
            else -> true
        }
    }

    /**
     * Check if a preference stores data as a JSON string
     * These preferences use ListGroupPreference which stores data as JSON-encoded strings
     */
    private fun isJsonStringPreference(key: String): Boolean {
        return when (key) {
            // LayoutsPreference - stores List<Layout> as JSON string
            "layouts",
            // ExtraKeysPreference - stores Map<KeyValue, PreferredPos> as JSON string
            "extra_keys",
            // CustomExtraKeysPreference - stores Map<KeyValue, PreferredPos> as JSON string
            "custom_extra_keys" -> true
            else -> false
        }
    }

    /**
     * Check if a preference is internal state that shouldn't be exported/imported
     */
    private fun isInternalPreference(key: String): Boolean {
        return when (key) {
            // Internal version tracking
            "version",
            // Current layout indices (managed by Config, device-specific)
            "current_layout_portrait",
            "current_layout_landscape" -> true
            else -> false
        }
    }

    /**
     * Check if a preference is stored as a float in SharedPreferences
     * This is critical because SharedPreferences throws ClassCastException if you
     * try to read an int as float or vice versa
     */
    private fun isFloatPreference(key: String): Boolean {
        return when (key) {
            // Character and UI sizing
            "character_size", "key_vertical_margin", "key_horizontal_margin", "custom_border_line_width",
            // Prediction weights
            "prediction_context_boost", "prediction_frequency_scale",
            // Auto-correction threshold
            "autocorrect_char_match_threshold",
            // Neural confidence threshold
            "neural_confidence_threshold",
            // Swipe typing boost parameters (SlideBarPreference floats)
            "swipe_rare_words_penalty", "swipe_common_words_boost", "swipe_top5000_boost" -> true
            else -> false
        }
    }

    /**
     * Check if a preference stores integers as strings (from ListPreference)
     * These need to be parsed and stored as int to prevent ClassCastException
     *
     * IMPORTANT: ListPreference always stores values as strings, even if they look like numbers.
     * Do NOT add ListPreference keys here - they must be imported as strings.
     */
    private fun isIntegerStoredAsString(key: String): Boolean {
        // Currently no preferences need this treatment
        // ListPreferences (show_numpad, circle_sensitivity, clipboard_history_limit) store as strings
        return false
    }

    /**
     * Check if a preference is stored as a StringSet
     * Prevents accidentally parsing other array types as StringSet
     */
    private fun isStringSetPreference(key: String): Boolean {
        // Currently no known StringSet preferences in this app
        // Add keys here if StringSet preferences are added in the future
        return false
    }

    /**
     * Validate string preference values
     */
    private fun validateStringPreference(key: String, value: String?): Boolean {
        if (value == null) return false

        return when (key) {
            // Theme values - relaxed validation for forward compatibility
            "theme" -> value.isNotEmpty()

            // Number row options
            "number_row" -> value.matches(Regex("no_number_row|no_symbols|symbols"))

            // Show numpad options
            "show_numpad" -> value.matches(Regex("never|always|landscape|[0-9]+"))

            // Numpad layout
            "numpad_layout" -> value.matches(Regex("high_first|low_first|default"))

            // Number entry layout
            "number_entry_layout" -> value.matches(Regex("pin|number"))

            // Circle sensitivity (string representation)
            "circle_sensitivity" -> value.matches(Regex("[1-5]"))

            // Slider sensitivity (string representation)
            "slider_sensitivity" -> value.matches(Regex("[0-9]+"))

            // Swipe distance (string representation)
            "swipe_dist" -> value.matches(Regex("[0-9]+(\\.[0-9]+)?"))

            // Unknown string preference - allow it (version-tolerant)
            else -> true
        }
    }

    /**
     * Result of import operation
     */
    data class ImportResult(
        @JvmField var importedCount: Int = 0,
        @JvmField var skippedCount: Int = 0,
        @JvmField var sourceVersion: String = "unknown",
        @JvmField var sourceScreenWidth: Int = 0,
        @JvmField var sourceScreenHeight: Int = 0,
        @JvmField var currentScreenWidth: Int = 0,
        @JvmField var currentScreenHeight: Int = 0,
        @JvmField val importedKeys: MutableSet<String> = mutableSetOf(),
        @JvmField val skippedKeys: MutableSet<String> = mutableSetOf()
    ) {
        fun hasScreenSizeMismatch(): Boolean {
            if (sourceScreenWidth == 0 || sourceScreenHeight == 0)
                return false // No source dimensions available

            val widthDiff = abs(currentScreenWidth - sourceScreenWidth)
            val heightDiff = abs(currentScreenHeight - sourceScreenHeight)

            // Consider it a mismatch if either dimension differs by more than 20%
            return (widthDiff > currentScreenWidth * 0.2) ||
                (heightDiff > currentScreenHeight * 0.2)
        }
    }

    companion object {
        private const val TAG = "BackupRestoreManager"
    }
}
